metriken-exposition/src/parquet.rs

```diff
@@ -0,0 +1,277 @@
+
+use crate::snapshot::{HashedSnapshot, Snapshot};
+use crate::HistogramSnapshot;
+use arrow::array::*;
+use arrow::datatypes::*;
+use arrow::error::ArrowError;
+use parquet::arrow::ArrowWriter;
+use parquet::basic::{Compression, ZstdLevel};
+use parquet::errors::ParquetError;
+use parquet::file::properties::WriterProperties;
+use parquet::format::FileMetaData;
+
+pub struct ParquetWriter {
+    /// File handle for output parquet file
+    file: Option<File>,
+    writer: Option<ArrowWriter<File>>,
+    properties: WriterProperties,
+
+    /// Number of rows cached in memory before being written as a `RecordBatch`
+    max_batch_size: usize,
+
+    /// Schema and columnar data for timestamps
+    tses: Vec<u64>,
+
+    /// Schema and columnar data for counters
+    counters: BTreeMap<String, Vec<Option<u64>>>,
+
+    /// Schema and columnar data for gauges
+    gauges: BTreeMap<String, Vec<Option<i64>>>,
+
+    /// Schema and columnar data for histograms
+    histograms: BTreeMap<String, Vec<Option<HistogramSnapshot>>>,
+
+    /// Schema generated for the parquet file after finalizing metadata
+    schema: Arc<Schema>,
+}
+
+/// Convert metrics snapshot data into parquet
+///
+/// Parquet files need the entire schema to be known prior to starting to
+/// write any data and all the columns to be of the same length. Missing
+/// data can be filled-up using NULL/empty values (columns can be marked
+/// nullable), but must be accounted for.
+///
+/// Metrics snapshot data does not have a well-defined schema, i.e., dynamic
+/// metrics may appear in some snapshots, but not in others, all the snapshots
+/// must be scanned and a complete list of all metrics extracted before
+/// the parquet file can be written. Further, since the parquet file may be
+/// too large to fit into memory, snapshots should be processed in batches.
+///
+/// This means that we need to make two passes over the snapshots:
+/// a) First, we collect the cumulative union of all the metrics that have
+///    been seen in any snapshot and build the schema.
+/// b) We process batches of snapshots extracing the row-centric data and
+///    converting it to a columnar form while filling in `None` values for
+///    metrics which are missing in specific snapshots.
+impl ParquetWriter {
+    const DEFAULT_MAX_BATCH_SIZE: usize = 1024 * 1024;
+
+    pub fn try_new(file: File, compression: bool) -> Result<Self, ParquetError> {
+        Self::try_new_with_max_batch_size(file, Self::DEFAULT_MAX_BATCH_SIZE, compression)
+    }
+
+    pub fn try_new_with_max_batch_size(
+        file: File,
+        max_batch_size: usize,
+        compression: bool,
+    ) -> Result<Self, ParquetError> {
+        let props = match compression {
+            false => WriterProperties::builder()
+                .set_compression(Compression::UNCOMPRESSED)
+                .build(),
+            true => WriterProperties::builder()
+                .set_compression(Compression::ZSTD(ZstdLevel::try_new(3)?))
```

Comment:
This doesn't need to be a result. You know compression level 3 will always be a valid level. Use an `expect` here.

As a bonus, this means the method doesn't need to return a result.